### ISSUE TITLE
Portable Class Library

### ISSUE THREAD

**Issue by DynamicDave** [LordMike/TMDbLib#44]
Dear LordMike,

Just an idea, but isn't it possible to convert the (excellent) TMDbLib project to a Portable Class Library (PCL). You would have to drop the RestSharp reference, but a PCL can make use of the ASP.NET Web API client to collect and parse JSON data.

Like I said... It's just an idea, but after the conversion the TMDbLib can be referenced by another PCL and be used in several project types (Win8, WinPhone8, etc).

Dave

**Comment by bbougot** [comment]
:+1: , really need it!

**Comment by LordMike** [comment]
@bbougot, I'm guessing it's for Mobile apps?

But if @Naliath agrees, perhaps we should move to MS-HttpClient and Newtonsoft.Json for v. 1.0.0?

**Comment by bbougot** [comment]
@LordMike Yup that's it! I'd like to use it into a Universal App project

**Comment by Naliath** [comment]
This seems the way most rest sharp users are moving in any case (checked some issues involving this exact request from the ResetSharp repo)

**Comment by DynamicDave** [comment]
I haven't been working on the PCL version for a while. Perhaps I can start working on the project again soon. But I have a lot of other work to do, so I'm lacking time.

**Comment by LordMike** [comment]
The issue is just that. Time.
But now that we're merging versions (async -> master), perhaps we can do
the next step: Remove Restsharp.

I've forgotten everything I know about PCL. Is it a new target, or just a
convention when no non-PCL libraries are referenced?
Is it a new Nuget target, how will it affect other users of Nuget builds,
etc.. ? :P

Mvh.
Michael

**Comment by DynamicDave** [comment]
A PCL is a Portable Class Library. MS says: "The Portable Class Library project enables you to write and build managed assemblies that work on more than one .NET Framework platform. You can create classes that contain code you wish to share across many projects, such as shared business logic, and then reference those classes from different types of projects."

It's basically a subset of the .Net framework. So not all functionality is available. You can reference it, for example, from a winforms project but also from a Xamarin, WinPhone or Xbox project.

**Comment by Naliath** [comment]
Seems it's a separate project type. So it would result in:
- Create new project
- Move all files over
- Replace all unsupported dependencies

Preferable we can just replace the project file so we can retain the file history on the code files

Time is always an issue but this should be relatively easy to implement since we would just be replacing something that is already centralized. If we wrap the new lib in a thin wrapper it should be OK.

**Comment by DynamicDave** [comment]
Dear Naliath, that's why I've made the separate branch
https://github.com/DynamicDave/TMDbLib

**Comment by Naliath** [comment]
@DynamicDave yea perfect just saying there have been quite a few changes since you created the branch so we will have to do some merging :)

**Comment by DynamicDave** [comment]
I know :-)

**Comment by LordMike** [comment]
I imagine it would be as simple as declaring that "henceforth we be PCL", and then fixing all compile errors. :P

Maybe we should move to HttpClient(Newtonsoft in a seperate move though.

**Comment by bbougot** [comment]
@LordMike There's a portable version of RestSharp here :
https://github.com/FubarDevelopment/restsharp.portable
It can be useful for the PCL version of TMDBLib

**Comment by LordMike** [comment]
Hmm, also published to Nuget: https://www.nuget.org/packages/FubarCoder.RestSharp.Portable/

**Comment by LordMike** [comment]
@DynamicDave we've just worked up a `RestSharp` replacemnet using `HttpClient`. Could you try this out?

Note that the library is now completely `async`, so you might have to change a few calls. :)

**Comment by Naliath** [comment]
Make a quick PCL version, right now just put a new project next to the old one. I didn't make any namespace changes so should be able to just dump the new over the old after changing the project name and project namespace

@LordMike You care to take a look? Just ran the test app and that looks fine, didn't check in more detail. Just felt like hacking something together :)

**Comment by LordMike** [comment]
Thought of doing the same. I'll try looking tomorrow. I imagine, for history's sake, we'll do a commit that simply changes the project type, instead of moving files :P.

-----Original Message-----

**Comment by LordMike** [comment]
I'm trying it out now. It seems very promising - running tests take time though.

**Comment by LordMike** [comment]
@Naliath looks really good. I've created issues for the two problems I found running tests. You already took care of the Description thing - but the HttpValueCollection I might rewrite to something smaller.

**Comment by Naliath** [comment]
Good job, yea the HttpValueCollection was just something I grabbed to get it to work and changed it only where there where issues getting things to build.

Besides the things you changed there where some issues with enums that I'm not sure I got a 100% replacement for. It seems PCL does not support the binding flags we used before to filter properties so it might be we are casting the net a bit wider than we were deserializing those props. I don't think it's an issue but still something to take note off.

**Comment by LordMike** [comment]
Yea I noticed. I think I got it covered though by changing the enum type check, and then only taking a field who's value is equal to what were looking for.

It wont work for flag types though. By then again, neither did the previous code.

Reflection is severely limited in pcl. Also, inspecting the code behind the extension method for get fields revealed a "return null". Reading up on it, it seems that its a way of exposing platform specific functionality in pcl (the platform dll delivers the implementation)... It should work in all places, but I fear it may fail on some obscure platform.

Alternatively, we'll just replace the entire thing with a switch/return static strings.. :P

-----Original Message-----
